### PR TITLE
Make Gitpod cookie "stricter"

### PR DESCRIPTION
**Pull request by AlexTugarev**
## Description
This PR changes the way we're managing [sameSite](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie/SameSite) setting of the Gitpod cookie (aka session cookie). 

We know, having them on `lax` mode in general comes with a reduction of browser based security mechanism. OTOH the current OAuth2 implementation relies on the session cookie being present on redirects. Also, during the OAuth2 flow, the mentioned browser security model might be neglected, at least there would be other measures possible to harden that process not relying on the browser agent. Given that, we can improve the security posture, by switching to `lax` mode for OAuth2 processes and back to `strict` mode when the process is done.

## Related Issue(s)

Fixes #

## How to test
1. Try to login with GitHub and GitLab.
2. Repeat after signing out of GitHub and GitLab to simulate a cold start of IdP session, which should lead to 

Screenshot 2023-02-15 at 10 07 13

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-github-actions
      Experimental feature to run the build with GitHub Actions (and not in Werft).
- [ ] leeway-no-cache
      leeway-target=components:all
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-ee-license
- [ ] with-slow-database
- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] /werft with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

### CODE REVIEW

**Comment by roboquat** [comment]
@AlexTugarev: Adding the "do-not-merge/release-note-label-needed" label because no release-note block was detected, please follow our [release note process](https://git.k8s.io/community/contributors/guide/release-notes.md) to remove it.

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository.
</details>

**Comment by AlexTugarev** [comment]
@geropl, please have a look.

File: components/server/src/auth/generic-auth-provider.ts [hunk 1]
```diff
@@ -145,11 +145,20 @@ export class GenericAuthProvider implements AuthProvider {
     protected readAuthUserSetup?: (accessToken: string, tokenResponse: object) => Promise<AuthUserSetup>;
 
     authorize(req: express.Request, res: express.Response, next: express.NextFunction, scope?: string[]): void {
-        const handler = passport.authenticate(this.getStrategy() as any, {
-            ...this.defaultStrategyOptions,
-            ...{ scope },
+        req.session.cookie.sameSite = "lax";
```
Comment (AlexTugarev) [inline components/server/src/auth/generic-auth-provider.ts:148]: setting to `lax` before redirecting to GH/GL

**Comment by geropl** [comment]
Ah, nice idea! :bulb: :sparkles: 

I wonder if it's worth the effort. But will definitely have a closer look after lunch. :pray:

File: components/server/src/auth/generic-auth-provider.ts [hunk 2]
```diff
@@ -466,6 +475,22 @@ export class GenericAuthProvider implements AuthProvider {
                     ...defaultLogPayload,
                 });
 
+                try {
+                    await new Promise((resolve, reject) => {
+                        request.session.cookie.sameSite = "strict";
```
Comment (AlexTugarev) [inline components/server/src/auth/generic-auth-provider.ts:480]: setting to `strict` on end of the auth process.

File: components/server/src/auth/generic-auth-provider.ts [hunk 3]
```diff
@@ -466,6 +475,22 @@ export class GenericAuthProvider implements AuthProvider {
                     ...defaultLogPayload,
                 });
 
+                try {
```
Comment (geropl) [inline components/server/src/auth/generic-auth-provider.ts:478]: A comment, or even a method, would be nice, to call out what the purpose of this piece of code is.

File: components/server/src/session-handler.ts [hunk 4]
```diff
@@ -53,7 +53,7 @@ export class SessionHandlerProvider {
             httpOnly: true, // default
             secure: false, // default, TODO SSL! Config proxy
             maxAge: config.session.maxAgeMs, // configured in Helm chart, defaults to 3 days.
-            sameSite: "lax", // default: true. "Lax" needed for OAuth.
+            sameSite: "strict", // default: true. "Lax" needed for OAuth.
```
Comment (AlexTugarev) [inline components/server/src/session-handler.ts:56]: TODO: update comment.

File: components/server/src/auth/generic-auth-provider.ts (hunk 1 above)
Comment (geropl) [reply components/server/src/auth/generic-auth-provider.ts:148]: :point_up: that should be part of a bigger comment explaining what we aim for, and pointing to the related parts of the code further down/in session-handler. :pray:

File: components/server/src/auth/generic-auth-provider.ts (hunk 3 above)
Comment (AlexTugarev) [reply components/server/src/auth/generic-auth-provider.ts:480]: done

File: components/server/src/auth/generic-auth-provider.ts (hunk 1 above)
Comment (AlexTugarev) [reply components/server/src/auth/generic-auth-provider.ts]: done

**Comment by AlexTugarev** [comment]
/hold 

I didn't test (re-)starting workspaces, thus it's not verified if that works with IDE at all.

**Comment by AlexTugarev** [comment]
/hold cancel

Starting workspaces ✔️ 
Restarting workspace (after logout in background) ✔️

**Pull request closed without merging**